Added docs for Valkey Search commands

eifrah-aws opened the pull request:
Like the title suggests: this PR adds the files + changes required for creating man pages for valkey-search module

yairgott commented on commands/ft.create.md:
```diff
@@ -0,0 +1,56 @@
+## FT.CREATE
```
Here and all the other commands, it would be useful to add examples. You can use the examples provided in Memorystore docs: https://cloud.google.com/memorystore/docs/redis/ftcreate#examples

Also, it would be useful to provide code samples, similar to: https://cloud.google.com/memorystore/docs/cluster/ftsearch#code_sample. But I'm not sure if Valkey documentation supports tabs.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,49 @@
+# Valkey-Search
```
Missing documentation on configurations and info fields.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,49 @@
+
+If replica lag is acceptable, users can achieve horizontal query scaling by directing clients to read from replicas.
+
+## Performance
```
This is more of a section for a blog, it doesn't really indicate much about how to use search functionality. The persona that is reading the website is looking for how to use search.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,49 @@
+
+## Hybrid Queries
+
+Valkey-Search supports hybrid queries, combining Vector Similarity Search with filtering on indexed fields, such as **Numeric** and **Tag indexes**.
```
I'm only familiar with tag indexes. Is this the comprehensive list? Is there documentation about the comprehensive list?

yairgott replied:
Beside Vector the only additional index types are Tag and Numeric :)

eifrah-aws replied:
I updated the doc + added examples and detailed explanation

madolson commented on topics/search.md:
````diff
@@ -0,0 +1,49 @@
+FT.SEARCH
+```
+
+For a detailed description of the supported commands and configuration options, see the [Command Reference](https://github.com/valkey-io/valkey-search/blob/main/COMMANDS.md).
````
Stay within the website, we should link to valkey.io/commands/#search presumably.

madolson commented on commands/ft._list.md:
```diff
@@ -0,0 +1,11 @@
+## FT._LIST
```
The command name is automatically added, you can omit this. You can omit the back ticks as well.

madolson commented on commands/ft._list.md:
```diff
@@ -0,0 +1,11 @@
+
+Lists the currently defined indexes.
+
+**RESPONSE**
+
+An array of strings which are the currently defined index names.
```
Should be generated from the .json files, you can include examples here, but the response type comes from the resp2/resp3 replies files.

eifrah-aws replied:
So we should not add the `**RESPONSE**` section? it will be auto generated by the CI?

madolson commented on groups.json:
```diff
@@ -7,6 +7,10 @@
   },
+  "search": {
+    "display": "Search",
+    "description": "Vector Similarity Search engine optimized for AI-driven workloads"
```
```suggestion
    "description": "Vector Similarity Search engine optimized for AI-driven workloads"
```
The intention is to make this a generic search module? We should probably not highlight just vectory search right now.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,49 @@
+# Valkey-Search
+
+**Valkey-Search** (BSD-3-Clause), provided as a Valkey module, is a high-performance Vector Similarity Search engine optimized for AI-driven workloads. It delivers single-digit millisecond latency and high QPS, capable of handling billions of vectors with over 99% recall.
```
We don't have a format checker on this, but you should either put a newline after each sentence or wrap around 80-100 lines. This is so it's much easier to edit lines without having extremely long edits.

eifrah-aws replied:
I broke the lines to make it more "terminal" fitted

yairgott commented:
LGTM

yairgott commented on topics/search.md:
```diff
@@ -0,0 +1,157 @@
+
+For a detailed description of the supported commands, examples and configuration options, see the [Command Reference](/commands/#search).
+
+## Configuration
```
1. The list of the supported load-module configuration can be found [here](https://github.com/valkey-io/valkey-search/blob/main/src/valkey_search.cc#L84-L89). 
2. We should distinguish between static load-module configuration and configuration which can be dynamically loaded. Today, we only support setting the hnsw [block size](https://github.com/valkey-io/valkey-search/blob/main/src/valkey_search.cc#L591) dynamically.

madolson commented on commands/ft.create.md:
````diff
@@ -0,0 +1,83 @@
+
+```
+FT.CREATE <index-name>
+    ON HASH | JSON
````
```suggestion
    [ON HASH | JSON]
```
You say these are optional. This will also automatically be generated based on the json files in the search repository, so do we need a secondary example here?

madolson commented on commands/ft.dropindex.md:
````diff
@@ -0,0 +1,8 @@
+```bash
+FT.DROPINDEX <index-name>
+```
+
````
```suggestion
```

madolson commented on commands/ft.info.md:
````diff
@@ -0,0 +1,36 @@
+```
+FT.INFO <index-name>
+```
+
````
```suggestion
```

madolson commented on commands/ft.create.md:
````diff
@@ -0,0 +1,83 @@
+**HNSW example:**
+
+```
+FT.CREATE my_index_name SCHEMA my_hash_field_key VECTOR HNSW 10 TYPE FLOAT32 DIM 20 DISTANCE_METRIC COSINE M 4 EF_CONSTRUCTION 100
````
Example should include the return result.

madolson commented on commands/ft.search.md:
````diff
@@ -0,0 +1,235 @@
+```
+FT.SEARCH <index> <query>
+  [NOCONTENT]
+  [TIMEOUT <timeout>]
+  [PARAMS nargs <name> <value> [ <name> <value> ...]]
+  [LIMIT <offset> <num>]
+  [DIALECT <dialect>]
+```
````
```suggestion
```

madolson commented on resp2_replies.json:
```diff
@@ -461,6 +461,23 @@
   ],
+  "FT.CREATE": [
+    "[Simple string reply](../topics/protocol.md#simple-strings): `OK`.",
+    "",
```
When can this return empty?

madolson commented on resp2_replies.json:
```diff
@@ -461,6 +461,23 @@
+  "FT.CREATE": [
+    "[Simple string reply](../topics/protocol.md#simple-strings): `OK`.",
+    "",
+    "If an error occurs during the index creation, an error is returned."
```
I don't think we explicitly call out error responses. I don't see another example anywhere.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+The following metrics are added to the `INFO` command's output:
+
+- **search_used_memory_human** (Integer) A human-friendly readable version of the `search_used_memory_bytes` metric
```
```suggestion
- `search_used_memory_human`: A human-friendly readable version of the `search_used_memory_bytes` metric
```
This is more closely related to the format we use for info fields in other places like bloom https://raw.githubusercontent.com/valkey-io/valkey-doc/refs/heads/main/topics/bloomfilters.md or the info command.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+- **search_hnsw_remove_exceptions_count** (Integer) Count of HNSW removal exceptions.
+- **search_hnsw_add_exceptions_count** (Integer) Count of HNSW addition exceptions.
+- **search_hnsw_modify_exceptions_count** (Integer) Count of HNSW modification exceptions
+- **search_modify_subscription_skipped_count** (Integer) Count of skipped subscription modifications
+- **search_remove_subscription_successful_count** (Integer) Count of successful subscription removals
+- **search_remove_subscription_skipped_count** (Integer) Count of skipped subscription removals
+- **search_remove_subscription_failure_count** (Integer) Count of failed subscription removals
+- **search_add_subscription_successful_count** (Integer) Count of successfully added subscriptions
+- **search_add_subscription_failure_count** (Integer) Count of failures of adding subscriptions
+- **search_add_subscription_skipped_count** (Integer) Count of skipped subscription adding processes
+- **search_modify_subscription_failure_count** (Integer) Count of failed subscription modifications
+- **search_modify_subscription_successful_count** (Integer)	Count of successful subscription modifications
```
I still have no idea what any of these are. Aren't these debug metrics that are only get included if you explicitly build for them?

eifrah-aws replied:
@yairgott

madolson commented on wordlist:
```diff
@@ -1013,3 +1013,29 @@ SlackBuilds
+paren
+tokenized
+pre
+search_hnsw_create_exceptions_count
+search_hnsw_search_exceptions_count
+search_hnsw_remove_exceptions_count
+search_hnsw_add_exceptions_count
+search_hnsw_modify_exceptions_count
```
If you use backticks, then these won't get flagged as mispelled words.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+- **search_background_indexing_status** (String) The status of the indexing process. `NO_ACTIVITY` indicates idle indexing.
+- **search_failure_requests_count** (Integer) A count of all failed requests, including syntax errors.
+- **search_successful_requests_count** (Integer) A count of all successful requests
+- **search_hnsw_create_exceptions_count** (Integer) Count of HNSW creation exceptions.
+- **search_hnsw_search_exceptions_count** (Integer) Count of HNSW search exceptions
+- **search_hnsw_remove_exceptions_count** (Integer) Count of HNSW removal exceptions.
```
What is the definition of an exception? We almost already refer to commands as errors everywhere else in the code base.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+Valkey-Search uses a **hybrid approach** with a query planner that selects the most efficient query execution path between:
+
+- **Pre-filtering**
```
Seems like you're missing some text here. You only define inline filtering. Also I'm confused, there is pre-filtering, post-filtering, and inline filtering?

eifrah-aws replied:
Fixed

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+- Empty strings are neither indexed or queried.
+- During indexing and querying, any trailing whitespace is removed.
+
+**Syntax**
```
```suggestion
#### Syntax
```

madolson replied:
This also needs some form of introduction.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+1. **--reader-threads:** (Integer) Controls the amount of threads executing queries.
+2. **--writer-threads:** (Integer) Controls the amount of threads processing index mutations.
+3. **--use-coordinator:** (boolean) Cluster mode enabler.
+4. **--hnsw-block-size:** (Integer) Controls the HNSW index resize increment steps.
```
Block size seems to be in the static and runtime section?

eifrah-aws replied:
Correct. If my PR: https://github.com/valkey-io/valkey-search/pull/129 will make it to 1.0 than `--reader-threads` & `--writer-threads` will also be runtime configurations

madolson commented on commands/ft.search.md:
```diff
@@ -0,0 +1,235 @@
+
+**RESPONSE**
+
+The command returns either an array if successful or an error.
```
This will get printed out twice, once from the replies section that is generated and here. I think we can omit most of this still.

madolson commented on commands/ft.search.md:
```diff
@@ -0,0 +1,235 @@
+  - **AS** This keyword is accompanied by a string value which becomes the name of the score field in the result,
+    overriding the default score field name generation algorithm.
+
+**Filter Expression**
```
```suggestion
 ### Filter Expression
```
I'm not entirely sure what header level these should all be in, but we should be using headers instead of bolding stuff. This applies to this entire file.

eifrah-aws replied:
Fixed

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+For a detailed description of the supported commands, examples and configuration options, see the [Command Reference](/commands/#search).
+
+## Configuration
```
Configuration should come later, what VSS is and how to use it should come first.

madolson commented on commands/ft.search.md:
```diff
@@ -0,0 +1,235 @@
+
+`-@genre:[comedy] @year:[2015 2024]`
+
+## Complete example: Simple vector search query
```
Really like this!

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+- **search_used_memory_bytes** (Integer) The total bytes of memory that all indexes occupy
+- **search_number_of_indexes** (Integer) Index schema total count
+- **search_number_of_attributes** (Integer) Total count of attributes for all indexes
+- **search_total_indexed_hash_keys** (Integer) Total count of HASH keys for all indexes
```
@yairgott Is there a reason there is no "indexed_json_keys" ?

yairgott replied:
Good catch! This is actually a naming bug, this counter captures the total keys of both hash and json data-types. Lets rename it to `search_total_indexed_documents`. I'll make this change in valkey-search.

yairgott replied:
See [fix](https://github.com/valkey-io/valkey-search/pull/137)

madolson reviewed (commented):
A lot of stuff is **bolded** that should probably be `backticked` to be consistent with the rest of our documentation. Overall the content seems OK to me, some minor suggestions for structure.

I think we need a better "getting started guide" for search, for someone who doesn't know anything to understand how to use it. I'm OK with that coming later.

eifrah-aws commented:
> A lot of stuff is **bolded** that should probably be `backticked` to be consistent with the rest of our documentation. Overall the content seems OK to me, some minor suggestions for structure.
> 
> I think we need a better "getting started guide" for search, for someone who doesn't know anything to understand how to use it. I'm OK with that coming later.

I too noticed it and changed it to backtick. However, the bold stuff looks much better in the man pages, so I reverted it back to **bold**. (same source is used for both the man pages and the website). If it is a blocker, I can switch it to backticks

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+# Valkey-Search
+
+**Valkey-Search** (BSD-3-Clause), provided as a Valkey module, is a high-performance Vector Similarity Search engine
```
```suggestion
**Valkey-Search** (BSD-3-Clause), provided as a Valkey module, is a high-performance vector similarity search engine
```
Some odd capitalization?

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+Users can index data using either **Valkey Hash** or **[Valkey-JSON](/topics/valkey-json)** data types.
+
+While Valkey-Search currently focuses on Vector Search, its goal is to extend Valkey into a full-fledged search engine,
+supporting Full Text Search and additional indexing options.
```
```suggestion
supporting full-text search and additional indexing options.
```

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+The following list of configurations can be passed to the `loadmodule` command:
+
+1. **--reader-threads:** (Integer) Controls the amount of threads executing queries.
```
```suggestion
1. **--search.reader-threads:** (Integer) Controls the amount of threads executing queries.
```
I believe these still need to include the prefix.

eifrah-aws replied:
No. When using the command line version, there is prefix `search.`

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+1. **--reader-threads:** (Integer) Controls the amount of threads executing queries.
+2. **--writer-threads:** (Integer) Controls the amount of threads processing index mutations.
+3. **--use-coordinator:** (boolean) Cluster mode enabler.
```
```suggestion
3. **--use-coordinator:** (boolean) Cluster mode enabler.
```
I still am not clear if this has to be the "same" value as with `--cluster-enabled` or there are reasons this would be different.

eifrah-aws replied:
We can probably deduce this from the current cluster state. I will consult the team

madolson replied:
Ok, we can say this isn't a blocker and can update it later.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+The following list of configurations can be modified at runtime using the `CONFIG SET` command:
+
+1. **search.hnsw-block-size:** (Integer) Controls the HNSW index resize increment steps.
```
Ideally *how* you would set this value should be included.

eifrah-aws replied:
I am not sure I follow... can you please clarify?

madolson replied:
Basically, what is the end user supposed to about this value. When would they increase it from the default, when would they decrease it.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,167 @@
+
+- Only indexed fields can be used as a tag filter.
+- TAG fields are tokenized by a separator character, which is a comma "," by default but configurable during index creation.
+- No stemming is performed while indexing a tag field.
```
What is stemming? This is only mentioned here in the entire doc.

yairgott replied:
Stemming is a text processing technique that reduces words to their root form. For example:

"running", "runner", and "ran" -> "run"

madolson commented on commands/ft.info.md:
```diff
@@ -0,0 +1,36 @@
+- **num\_records**	(integer)	Total records in the index
+- **hash\_indexing\_failures**	(integer)	Count of unsuccessful indexing attempts
+- **indexing**	(integer)	Binary value. Shows if background indexing is running or not
+- **percent\_indexed**	(integer)	Progress of background indexing. Percentage is expressed as a value from `0` to `1`
```
```suggestion
- **percent\_indexed**	(integer)	Progress of background indexing. Percentage is expressed as a value from `0` to `1`
```
I suppose this isn't an integer, it's a decimal value?

madolson commented on commands/ft.info.md:
```diff
@@ -0,0 +1,36 @@
+- **indexing**	(integer)	Binary value. Shows if background indexing is running or not
+- **percent\_indexed**	(integer)	Progress of background indexing. Percentage is expressed as a value from `0` to `1`
+- **index\_definition**	(array)	An array of values defining the index
+  - **key\_type**	(string)	`HASH`. This is the only available key type.
```
```suggestion
  - **key\_type**	(string)	`HASH` or `JSON`.
```
Presumably

madolson reviewed (commented):
Read through it again trying to act like I didn't know what VSS was.

yairgott commented on topics/search.md:
```diff
@@ -0,0 +1,177 @@
+
+The following list of configurations can be passed to the `loadmodule` command:
+
+1. `--reader-threads`: Controls the amount of threads executing queries.
```
We should state that, unless explicitly configured, the number of reader and writer threads is automatically set to the number of physical CPU cores on the host machine.

madolson replied:
Good call out! Does that mean the number of threads could be greater than the number of cores? Is there a way to set affinity on the cores so they don't contend with the main thread?

eifrah-aws replied:
Done

yairgott commented on topics/search.md:
```diff
@@ -0,0 +1,177 @@
+
+The following list of configurations can be modified at runtime using the `CONFIG SET` command:
+
+1. `search.hnsw-block-size:`: Controls the HNSW index resize increment steps.
```
Suggest:
`search.hnsw-block-size:` (default: 10KB): Specifies the allocation block size used by the HNSW graph for storing new vectors. Larger block sizes may improve performance by enhancing CPU cache efficiency, but come at the cost of increased memory usage due to pre-allocation for potential future growth.

madolson replied:
Also, let's remove it from the startup configuration.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,176 @@
+- `search_hnsw_remove_exceptions_count`: Count of HNSW removal exceptions.
+- `search_hnsw_add_exceptions_count`: Count of HNSW addition exceptions.
+- `search_hnsw_modify_exceptions_count`: Count of HNSW modification exceptions
+- `search_modify_subscription_skipped_count`: Count of skipped subscription modifications
+- `search_remove_subscription_successful_count`: Count of successful subscription removals
+- `search_remove_subscription_skipped_count`: Count of skipped subscription removals
+- `search_remove_subscription_failure_count`: Count of failed subscription removals
+- `search_add_subscription_successful_count`: Count of successfully added subscriptions
+- `search_add_subscription_failure_count`: Count of failures of adding subscriptions
+- `search_add_subscription_skipped_count`: Count of skipped subscription adding processes
+- `search_modify_subscription_failure_count`: Count of failed subscription modifications
+- `search_modify_subscription_successful_count`: Count of successful subscription modifications
```
@yairgott Can you confirm these are only enabled in debug builds?

yairgott replied:
No, these are enabled also in non-debug. These are useful to get a signal if you may have set the index prefixes, at index creation, correctly.

madolson commented on topics/search.md:
```diff
@@ -0,0 +1,176 @@
+- `search_background_indexing_status` (String) The status of the indexing process. `NO_ACTIVITY` indicates idle indexing.
+- `search_failure_requests_count`: A count of all failed requests, including syntax errors.
+- `search_successful_requests_count`: A count of all successful requests
+- `search_hnsw_create_exceptions_count`: Count of HNSW creation exceptions.
+- `search_hnsw_search_exceptions_count`: Count of HNSW search exceptions
+- `search_hnsw_remove_exceptions_count`: Count of HNSW removal exceptions.
+- `search_hnsw_add_exceptions_count`: Count of HNSW addition exceptions.
+- `search_hnsw_modify_exceptions_count`: Count of HNSW modification exceptions
```
@yairgott Are exceptions the same as errors? My mental model is that an exception is an unhandled internal failure, but I don't know what that would map to here.

yairgott commented:
> Does that mean the number of threads could be greater than the number of cores? 
Yes
> Is there a way to set affinity on the cores so they don't contend with the main thread?
No

madolson reviewed (approved):
LGTM, thanks for all your hard work on this! @yairgott Can you also sign off to make sure everything is accurate?

eifrah-aws commented on commands/ft.dropindex.md:
```diff
@@ -0,0 +1,4 @@
+Drop the index created earlier by `FT.CREATE` command. It is an error if the index doesn't exist.
+
+- **\<index-name\>** (required): The name of the index to delete.
```
Done
